pkg/operator/encryption/shared_encryption.go

```diff
@@ -0,0 +1,431 @@
+			return grKeys.secretToKey[writeYes.Name], true
+		}
+	}
+
+	// first read that is not write
+	for _, readYes := range grKeys.secretsReadYes {
+		if len(readYes.Annotations[encryptionSecretWriteTimestamp]) == 0 {
+			return grKeys.secretToKey[readYes.Name], true
+		}
+	}
+
+	// no key is transitioning so just use last migrated
+	if len(grKeys.secretsMigratedYes) > 0 {
+		lastMigrated := grKeys.secretsMigratedYes[len(grKeys.secretsMigratedYes)-1]
+		return grKeys.secretToKey[lastMigrated.Name], true
+	}
+
+	// no write key
+	return apiserverconfigv1.Key{}, false
+}
+
+func keysToProviders(grKeys keysState) []apiserverconfigv1.ProviderConfiguration {
+	desired := grKeysToDesiredKeys(grKeys)
+
+	allKeys := desired.readKeys
+
+	// write key comes first
+	if desired.hasWriteKey {
+		allKeys = append([]apiserverconfigv1.Key{desired.writeKey}, allKeys...)
+	}
+
+	aescbc := apiserverconfigv1.ProviderConfiguration{
+		AESCBC: &apiserverconfigv1.AESConfiguration{
+			Keys: allKeys,
+		},
+	}
+	identity := apiserverconfigv1.ProviderConfiguration{
+		Identity: &apiserverconfigv1.IdentityConfiguration{},
+	}
+
+	// assume the common case of having a write key so identity comes last
+	providers := []apiserverconfigv1.ProviderConfiguration{aescbc, identity}
+	// if we have no write key, identity comes first
+	if !desired.hasWriteKey {
+		providers = []apiserverconfigv1.ProviderConfiguration{identity, aescbc}
+	}
+
+	return providers
+}
+
+func shouldRunEncryptionController(operatorClient operatorv1helpers.StaticPodOperatorClient) (bool, error) {
+	operatorSpec, _, _, err := operatorClient.GetStaticPodOperatorState()
+	if err != nil {
+		return false, err
+	}
+
+	return management.IsOperatorManaged(operatorSpec.ManagementState), nil
+}
+
+func getAPIServerRevision(podClient corev1client.PodInterface) (string, error) {
+	// do a live list so we never get confused about what revision we are on
+	apiServerPods, err := podClient.List(metav1.ListOptions{LabelSelector: "apiserver=true"})
+	if err != nil {
+		return "", err
+	}
+
+	revisions := sets.NewString()
+	for _, apiServerPod := range apiServerPods.Items {
+		switch phase := apiServerPod.Status.Phase; phase {
+		case corev1.PodRunning: // TODO check that total running == number of masters?
+			if !isPodReady(apiServerPod) {
+				return "", nil // pods are not fully ready
+			}
+			revisions.Insert(apiServerPod.Labels[revisionLabel])
+		case corev1.PodPending:
+			return "", nil // pods are not fully ready
+		case corev1.PodUnknown:
+			return "", fmt.Errorf("api server pod %s in unknown phase", apiServerPod.Name)
+		case corev1.PodSucceeded, corev1.PodFailed:
+			// old pods that we can safely ignore
+		default:
+			// do not error in case new phases get added
+			klog.Infof("api server pod %s has unexpected phase %v", apiServerPod.Name, phase)
+		}
+	}
+
+	if len(revisions) != 1 {
+		return "", nil // api servers have not converged onto a single revision
```

Comment:
we could at least log that we are waiting for pods to converge, otherwise, it could be hard to tell why the controller isn't running.